bot/src/bot.js

```diff
@@ -1,99 +1,97 @@
+export class Bot {
+  constructor(url, name, api_url) {
+    this.url = url;
+    this.name = name;
+    this.api_url = api_url;
+    this.commands = Commands;
+    this.actionLog = new Map();
+  }
 
-  client.addListener('invite', function(channel) {
-    console.log(` --- got an invite to ${channel}`);
+  start() {
+    this.client = new Client(this.url, this.name);
 
-    client.join(channel, () => client.say(channel, 'hello'));
-  });
+    this.client.addListener('invite', (channel) => {
+      console.log(` --- got an invite to ${channel}`);
 
-  // listen to messages in public channels
-  client.addListener('message#', function(author, channel, message) {
-    console.log(` --- got a message from ${author} in ${channel}: ${message}`);
+      this.client.join(channel, () => this.client.say(channel, 'hello'));
+    });
 
-    if (!message.startsWith(`${client.nick}: `)) {
-      // the message wasn't directed at the bot
-      return;
-    }
+    // listen to messages in public channels
+    this.client.addListener('message#', (author, channel, message) => {
+      console.log(` --- got a message from ${author} in ${channel}: ${message}`);
 
-    const text = message.slice(client.nick.length + 2);
-    parseCommand(db, author, channel, text).then(
-      response => client.say(channel, `${author}: ${response}`)
-    );
-  });
+      if (!message.startsWith(`${this.client.nick}: `)) {
+        // the message wasn't directed at the bot
+        return;
+      }
 
-  // listen to private messages
-  client.addListener('pm', function(author, message) {
-    console.log(` --- got a private message from ${author}: ${message}`);
+      const text = message.slice(this.client.nick.length + 2);
+      parseCommand(this, author, channel, text).then(
+        response => this.client.say(channel, `${author}: ${response}`)
+      );
+    });
 
-    // private messages are saved with channel = author
-    parseCommand(db, author, author, message).then(
-      response => client.say(author, response)
-    );
-  });
-}
+    // listen to private messages
+    this.client.addListener('pm', (author, message) => {
+      console.log(` --- got a private message from ${author}: ${message}`);
 
-function parseCommand(db, author, channel, message) {
-  if (message.startsWith('create a report for')) {
-    return createReport(db, message);
+      // private messages are saved with channel = author
+      parseCommand(this, author, author, message).then(
+        response => this.client.say(author, response)
+      );
+    });
   }
 
-  return saveUpdate(db, author, channel, message);
+  getActionLog(user, channel) {
+    const hash = `${user}-${channel}`;
+    if (!this.actionLog.has(hash)) {
+      return [];
+    }
+    return this.actionLog.get(hash);
+  }
+
+  logAction(user, channel, action) {
+    const ts = Date.now();
+    const hash = `${user}-${channel}`;
+    if (!this.actionLog.has(hash)) {
+      this.actionLog.set(hash, []);
+    }
```

Comment:
yeah, I know, I thought that maybe that was okay since it's not a common scenario, but on second thought it's probably better to keep it the way you wrote it right now. thanks!